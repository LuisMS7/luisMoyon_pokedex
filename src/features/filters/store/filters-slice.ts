import { create } from "zustand";
import { Filters } from "@filters/types/filters.type";

type State = {
	pokemonName: string;
	setPokemonName: (pokemonName: string) => void;
	filters: Filters;
	filterPokemons: () => void;
};

export const useFilters = create<State>((set, get) => ({
	pokemonName: "",
	setPokemonName: (pokemonName: string) => {
		set({
			pokemonName: pokemonName.toLowerCase().trim(),
		});
	},
	filters: { pokemonName: "" },
	filterPokemons: () => {
		set({
			filters: { pokemonName: get().pokemonName },
		});
	},
}));
