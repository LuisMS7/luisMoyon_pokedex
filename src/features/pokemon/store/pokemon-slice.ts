import { create } from "zustand";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";

type State = {
	pokemons: PokemonListItemApiResponse[];
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	clearPokemons: () => void;
};

const usePokemonStore = create<State>((set, get) => ({
	pokemons: [],
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => {
		const currentPokemonsList = get().pokemons;
		const updatePokemonsList = [...currentPokemonsList, ...pokemons];
		set({
			pokemons: updatePokemonsList,
		});
	},
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => {
		set({
			pokemons: pokemons,
		});
	},
	clearPokemons: () =>
		set({
			pokemons: [],
		}),
}));

export default usePokemonStore;
