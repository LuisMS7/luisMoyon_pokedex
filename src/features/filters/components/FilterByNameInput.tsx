import React from "react";
import { StyleSheet, Text, TextInput, View } from "react-native";
import {
	BOLD_WEIGHT,
	NORMAL_WEIGHT,
	TEXT_NORMAL_SIZE,
	TEXT_SMALL_SIZE,
} from "@lib/constants";
import { useFilters } from "@filters/store/filters-slice";

const FilterByNameInput = () => {
	const pokemonName = useFilters((state) => state.pokemonName);
	const setPokemonName = useFilters((state) => state.setPokemonName);

	const handleChangeText = (inputText: string) => {
		setPokemonName(inputText);
	};

	return (
		<View style={styles.container}>
			<Text style={styles.title}>Search</Text>
			<TextInput
				placeholder="Pokemon name"
				value={pokemonName}
				onChangeText={handleChangeText}
				style={styles.textInput}
				autoCapitalize="none"
				autoCorrect={false}
				testID="pokemon-name-text-input"
			/>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		justifyContent: "space-around",
		height: "100%",
	},
	title: {
		fontWeight: BOLD_WEIGHT,
		fontSize: TEXT_SMALL_SIZE,
	},
	textInput: {
		fontWeight: NORMAL_WEIGHT,
		fontSize: TEXT_NORMAL_SIZE,
	},
});

export default FilterByNameInput;
