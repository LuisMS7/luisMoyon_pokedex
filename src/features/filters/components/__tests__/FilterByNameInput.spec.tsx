import React from "react";
import { act, fireEvent, render, screen } from "@testing-library/react-native";
import FilterByNameInput from "@filters/components/FilterByNameInput";
import { useFiltersStore } from "@filters/store/filters-slice";

jest.mock("@filters/store/filters-slice", () => ({
	useFiltersStore: jest.fn(),
}));

describe("FilterByNameInput", () => {
	it("should render correctly", () => {
		(useFiltersStore as unknown as jest.Mock).mockImplementation(
			(selector) =>
				selector({ pokemonName: "", setPokemonName: jest.fn() }),
		);
		render(<FilterByNameInput />);

		expect(screen.toJSON()).toMatchSnapshot();
	});

	it("should call setPokemonName when user edit text input", () => {
		const mockSetPokemonName = jest.fn();
		(useFiltersStore as unknown as jest.Mock).mockImplementation(
			(selector) =>
				selector({
					pokemonName: "",
					setPokemonName: mockSetPokemonName,
				}),
		);
		const { getByTestId } = render(<FilterByNameInput />);

		const textInput = getByTestId("pokemon-name-text-input");

		act(() => {
			fireEvent.changeText(textInput, "pikachu");
		});

		expect(mockSetPokemonName).toHaveBeenCalledWith("pikachu");
	});

	it("should use pokemonName of store in text input", () => {
		(useFiltersStore as unknown as jest.Mock).mockImplementation(
			(selector) =>
				selector({
					pokemonName: "mockPokemon",
					setPokemonName: jest.fn(),
				}),
		);
		const { getByTestId } = render(<FilterByNameInput />);

		const textInput = getByTestId("pokemon-name-text-input");

		expect(textInput.props.value).toBe("mockPokemon");
	});
});
