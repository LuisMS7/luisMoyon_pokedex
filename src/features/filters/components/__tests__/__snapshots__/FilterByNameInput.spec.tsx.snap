// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByNameInput should render correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "justifyContent": "space-around",
    }
  }
>
  <Text
    style={
      {
        "fontSize": 17.89194962785868,
        "fontWeight": "800",
        "paddingTop": 4,
      }
    }
  >
    Search
  </Text>
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    onChangeText={[Function]}
    placeholder="Pokemon name"
    style={
      {
        "flex": 1,
        "fontSize": 20.873941232501792,
        "fontWeight": "400",
      }
    }
    testID="pokemon-name-text-input"
    value=""
  />
</View>
`;
